Chat

the-itz opened the pull request.

the-itz committed:
add Markdown, Sanitize-htm and Date-fns

the-itz committed:
add chat-message component

the-itz committed:
add chat-input-form

the-itz committed:
change html tags add formatting

the-itz committed:
add Edit Message

the-itz committed:
merge master

the-itz committed:
add cancel button and emojis

the-itz committed:
Merge branch 'master' into chat

the-itz committed:
add dropdown for @ mention

the-itz committed:
add reply, simplify dependencies

the-itz committed:
remove consol outputs

the-itz committed:
rename onSubmit to onClickSubmit

the-itz committed:
remove comment

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+}
+
+export default function ChatInputForm({ onCreate, onCancel, setInput, preSetInput, users, allowEscape }: Props) {
+  const [userInput, setUserInput] = useState(preSetInput);
```
I'd recommend to rename `userInput` to just `input`, `setUserInput` to `setInput`

the-itz replied:
Not sure about this. I have userInput for the what the user inputs and
preSetInput the text which the input form should display when there ist already something to edit
and setInput is the callback which is used  to trigger saving to the db + rerender.
The latter one would clash with your proposal. 
If you want to have the change. I would then rename the callback to setInputClbk

the-itz replied:
The thumb up means, you want me to change names? @arminfro

arminfro replied:
it means you're right, leave it like it is. I didn't saw the prop with the name `setInput`

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+  };
+
+  const selectAtMention = (index: number, value: string): void => {
+
```
this new line is not needed

the-itz replied:
Done

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+
+  const selectAtMention = (index: number, value: string): void => {
+
+    setUserInput(oldMsg + '**@' + value + '** ');
```
```suggestion
    setUserInput(oldMsg + '**@' + value + '** ');
```
```suggestion
    setUserInput(oldMsg + ' **@' + value + '** ');
```
Having a space before `**@...` makes sense, I guess.
Edit: Maybe we should check if there is a space before, if not insert one

the-itz replied:
done, incl check
Found a new issue there. We have to define the behaviour if the user presses @ in the middle of the text. Currently we just add the string to the end of the message. I'll create an issue for this.

the-itz replied:
is used the ternary operator here, didn't work with && though
Also tried to have ist inline in the setUserInput call -> messed up result. So preset const string used this to assemble result string.

the-itz replied:
Ok, learned something new. If you create an issue out of a comment, the comment will then disappear.
Done and issues created.

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+          onKeyDown={onKeyDown}
+          placeholder="Your Message"
+          id="msg"
+        ></textarea>
```
this element has no childs, can be closed immediately

the-itz replied:
done

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+          placeholder="Your Message"
+          id="msg"
+        ></textarea>
+        {doAtMention ? <Dropdown entries={reducedUserList.map((u) => u.name)} callback={selectAtMention} action={action} /> : ''}
```
Logical `&&` is sufficient, no ternary operator needed

the-itz replied:
done

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+        or <b>Ctrl-Return</b>
+      </span>
+
+      {allowEscape && <button onClick={onClickCancel} className="ui red labled submit icon button ">
```
`<button ...` should go on the next line, to keep indentation level of opening and closing tag

the-itz replied:
done

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+      </span>
+
+      {allowEscape && <button onClick={onClickCancel} className="ui red labled submit icon button ">
+        <i className="icon cancel"></i>cancel
```
indentation is wrong

the-itz replied:
Don't get this one, could you pls check the update of the file?

arminfro commented on src/pages/rooms/chat-input-form.tsx:
```diff
@@ -0,0 +1,130 @@
+import React, {
```
Many functions in this file have no return type specified

the-itz replied:
Done. Defined Return types of functions.

arminfro commented on src/pages/rooms/chat-message.tsx:
```diff
@@ -0,0 +1,113 @@
+
+  const createReplyString = (chat: Chat, newInput: string): string => {
+    return `Reply to ${chat.user.name}'s Message on ${new Date(chat.updatedAt).toLocaleDateString()}: \n ${chat.msg.replace(/^/gm, "> ")} \n\n  ${newInput}`
+  }
```
missing new line below

the-itz replied:
If you meant below  / after function, then done

arminfro commented on src/pages/rooms/chat-message.tsx:
```diff
@@ -0,0 +1,113 @@
+            {new Date(chat.updatedAt).getTime() !==
+              new Date(chat.createdAt).getTime()
+              ? '(edited)'
+              : ''}
```
Logical `&&` is sufficient, no ternary operator needed

the-itz replied:
Ok, done

arminfro commented on src/pages/rooms/chat-message.tsx:
```diff
@@ -0,0 +1,113 @@
+              </a>
+            </div>
+          )
+        ) : ('')
```
Logical `&&` is sufficient, no ternary operator needed

the-itz replied:
Done

arminfro commented on src/pages/rooms/chat-message.tsx:
```diff
@@ -0,0 +1,113 @@
+              : ''}
+          </span>
+        </div>
+        {(chat.links) ? (
```
return statement is quite long, maybe move this code into a new `ChatLink` component

arminfro commented on src/pages/rooms/chats.tsx:
```diff
@@ -27,30 +37,49 @@ export default function Chats({ roomId }: Props): ReactElement {
-    <div>
-      <div className="ui relaxed divided list">
+    <div className="ui container" style={{ maxHeight: '900px' }}>
+      <br></br>
```
this element has no childs, can be closed immediately

the-itz replied:
Done

arminfro commented on src/pages/rooms/dropdown.tsx:
```diff
@@ -0,0 +1,53 @@
+import React, { useEffect, useLayoutEffect, useRef, useState } from 'react';
+
+export interface Position {
```
is not used

the-itz replied:
Done

arminfro commented on src/pages/rooms/dropdown.tsx:
```diff
@@ -0,0 +1,53 @@
+}
+
+
+export default function Dropdown({ entries, callback, action }: Props) {
```
return type is missing

the-itz replied:
Done

arminfro commented on src/pages/rooms/dropdown.tsx:
```diff
@@ -0,0 +1,53 @@
+
+  useEffect(() => {
+    onAction(action)
+    return
```
`return` is not needed

the-itz replied:
Done

arminfro commented on src/pages/rooms/dropdown.tsx:
```diff
@@ -0,0 +1,53 @@
+
+  return (
+    <div className="dropdown" >
+      <input type="hidden" name="input"></input>
```
hidden input is needed for what?

the-itz replied:
Removed. Done.

arminfro commented on src/pages/rooms/dropdown.tsx:
```diff
@@ -0,0 +1,53 @@
+      <div className="dropdowntext" ref={dropdownRef}>
+        {entries.map((e, i) => {
+          return (
+            <div key={i} data-value={i} className={(selected == i) ? "selected" : ""} onClick={(ev) => select(i, e)}>
```
`data-value` is not needed, I think
Logical `&&` is sufficient, no ternary operator needed

the-itz replied:
Jepp not needed, removed and Done

the-itz committed:
implement Armin's comments

arminfro committed:
merge master into chat

adapt to chat tree structure

arminfro committed:
add editorconfig

the-itz commented on src/types/chat.ts:
```diff
@@ -17,6 +18,7 @@ export default interface Chat {
   readonly msg: string;
   readonly room: Room;
   readonly user: User;
+  readonly links: Link[];
```
@arminfro after the pull I've got this line twice. I'll remove the second one on my branch.

arminfro committed:
add some css classes to chat components

change html structure slightly

do soft-delete of chat

the-itz committed:
Cancel in no longer triggers reload
Issue #4

the-itz committed:
add breaks option for marked
layout improvements add spaces
add ago to time difference

the-itz committed:
solve issue #3

the-itz committed:
easier escaping from drop down while typing

the-itz committed:
dropdown no longer displayed if no entries exist

the-itz committed:
remove debug message

the-itz committed:
change hierarchy part 1
added ChatList, ChatListItem, ChatItemHeader, ChatLinkList
renamed ChatLink to ChatLinkListItem

the-itz committed:
Fix bug to prevent chat crash

the-itz committed:
update recursion to use only ChatList
remove chatMessage Component

the-itz committed:
remove import of ChatMessage

arminfro committed:
merge master, resolve conflicts

arminfro committed:
fix some linter issues

arminfro committed:
fix some more linter issues

revert some changes from 72e028a
